File: YukioPlayer/Sources/YukioPlayer/CardStack.swift
import AppKit
import YukioCore

/// 头顶那摞卡片里“别的聊天”那几张：一条聊天一张，叠在气泡上面，点一张就去那条聊天。
///
/// 外观就是头顶气泡那张卡（`CardLook`）：同样的圆角、描边、字号和两行排版，
/// 只多了一条状态色和一个 ✕。独立的透明子窗口，跟着雪绪走；
/// 只有卡片本身接收点击，卡与卡之间的缝隙照样穿透。
final class CardStackPanel: NSPanel {
    let stackView = CardStackView(frame: NSRect(x: 0, y: 0, width: 10, height: 10))

    init() {
        super.init(contentRect: NSRect(x: 0, y: 0, width: 10, height: 10),
                   styleMask: [.borderless, .nonactivatingPanel], backing: .buffered, defer: false)
        isOpaque = false
        backgroundColor = .clear
        hasShadow = false
        level = .floating
        collectionBehavior = [.canJoinAllSpaces, .fullScreenAuxiliary, .ignoresCycle]
        hidesOnDeactivate = false
        isReleasedWhenClosed = false
        isMovable = false
        ignoresMouseEvents = true
        alphaValue = 0
        contentView = stackView
    }

    override var canBecomeKey: Bool { false }
    override var canBecomeMain: Bool { false }
}

final class CardStackView: NSView {
    var layout: CardStackLayout? {
        didSet { needsDisplay = true }
    }

    /// 点了第几张卡的正文（去那条聊天）。
    var onOpen: ((Int) -> Void)?
    /// 点了第几张卡的 ✕（只收起这张）。
    var onDismiss: ((Int) -> Void)?
    /// 点了“还有 N 条 / 收起”。
    var onToggleExpand: (() -> Void)?
    /// 点了“自动：谁要紧跟谁”。
    var onPickAuto: (() -> Void)?
    /// 在第几张卡上右键（静音那条聊天）。
    var onContextMenu: ((Int, NSEvent) -> Void)?

    override var isOpaque: Bool { false }
    override func acceptsFirstMouse(for event: NSEvent?) -> Bool { true }

    override func draw(_ dirtyRect: NSRect) {
        layout?.draw(in: bounds)
    }

    /// 窗口内坐标（左下原点）上有没有可点的东西。
    func hit(at point: NSPoint) -> CardStackLayout.Hit? {
        layout?.hit(at: point, in: bounds)
    }

    override func mouseDown(with event: NSEvent) {
        // 和点雪绪一样，在 mouseUp 里处理。
    }

    override func mouseUp(with event: NSEvent) {
        switch hit(at: convert(event.locationInWindow, from: nil)) {
        case .card(let i): onOpen?(i)
        case .dismiss(let i): onDismiss?(i)
        case .auto: onPickAuto?()
        case .expand: onToggleExpand?()
        case nil: break
        }
    }

    override func rightMouseDown(with event: NSEvent) {
        if case .card(let i)? = hit(at: convert(event.locationInWindow, from: nil)) {
            onContextMenu?(i, event)
        }
    }
}

/// 这摞卡的排版与绘制，与窗口无关（`--cards` 快照也用它）。
///
/// 平时（收起）只有头顶那张气泡，这里画的仅仅是气泡上面一条“还有 N 条”的小条；
/// 点一下展开，才把别的聊天一条一张摊出来，每张点一下就把她换到那条聊天上。
/// 每张卡就是气泡那张卡的样子：上行聊天名（淡色小字），下行在做什么（深色粗字）。
struct CardStackLayout {
    enum Hit: Equatable {
        /// 选这条聊天当主题聊天。
        case card(Int)
        /// 只收起这一张提醒。
        case dismiss(Int)
        /// 回到自动（谁要紧跟谁）。
        case auto
        /// 展开／收起整摞。
        case expand
    }

    /// 和气泡同宽，叠起来是一摞齐的。
    static let width = CardLook.maxWidth
    /// 展开后最多摊几张。
    static let maxExpanded = 8
    private static let gap = CardLook.stackGap
    private static let stripe: CGFloat = 3
    private static let stripeGap: CGFloat = 5
    private static let closeBox: CGFloat = 15
    private static let rowHeight: CGFloat = 20
    /// 两条胶囊各自往里缩多少：画和命中用同一个数。
    private static let autoInset: CGFloat = 10
    private static let pillInset: CGFloat = 26

    static func color(for status: ActivityCard.Status) -> NSColor {
        switch status {
        case .waiting: return NSColor(srgbRed: 0.93, green: 0.62, blue: 0.16, alpha: 1)
        case .failed: return NSColor(srgbRed: 0.84, green: 0.31, blue: 0.29, alpha: 1)
        case .ready: return NSColor(srgbRed: 0.22, green: 0.68, blue: 0.44, alpha: 1)
        case .running: return CardLook.accent
        }
    }

    /// 摊出来的卡（收起时是空的）。
    let shown: [ActivityCard]
    /// 收起时折着的张数。
    let hidden: Int
    let expanded: Bool
    /// 展开时是否给一条“回到自动”。
    let showsAuto: Bool
    let size: NSSize
    /// 每张卡的矩形（左下原点，第 0 张在最下面、挨着气泡）。
    private let cardRects: [NSRect]
    private let autoRect: NSRect?
    private let pillRect: NSRect?

    /// maxHeight：气泡上面还剩多少地方。摊开后放不下就少摊几张（先扔最不要紧的，
    /// 也就是离气泡最远那几张），绝不压到气泡上。
    init(cards: [ActivityCard], expanded: Bool, pinned: Bool = false, maxHeight: CGFloat = .infinity) {
        self.expanded = expanded
        showsAuto = expanded && pinned
        let cardHeight = 2 * CardLook.padY + CardLook.lineHeight(CardLook.titleFont)
            + CardLook.lineGap + CardLook.lineHeight(CardLook.currentFont)
        var room = Self.maxExpanded
        if expanded, maxHeight.isFinite {
            var left = maxHeight - Self.rowHeight                      // “收起”那条
            if showsAuto { left -= Self.rowHeight + Self.gap }
            room = max(1, Int(floor((left + Self.gap) / (cardHeight + Self.gap))))
        }
        shown = expanded ? Array(cards.prefix(min(Self.maxExpanded, room))) : []
        hidden = cards.count - shown.count

        var rects: [NSRect] = []
        var y: CGFloat = 0
        for _ in shown {
            rects.append(NSRect(x: 0, y: y, width: Self.width, height: cardHeight))
            y += cardHeight + Self.gap
        }
        cardRects = rects
        if showsAuto {
            autoRect = NSRect(x: 0, y: y, width: Self.width, height: Self.rowHeight)
            y += Self.rowHeight + Self.gap
        } else {
            autoRect = nil
        }
        // 收起时只剩这一条；展开时它是“收起”。没有别的聊天就整个不画。
        if expanded || hidden > 0 {
            pillRect = NSRect(x: 0, y: y, width: Self.width, height: Self.rowHeight)
            y += Self.rowHeight
        } else {
            pillRect = nil
            if y > 0 { y -= Self.gap }
        }
        size = NSSize(width: Self.width, height: max(0, ceil(y)))
    }

    /// 这摞卡的左下角：和气泡同一竖线上居中，压在气泡上面往上长。
    /// bubbleTop 是气泡的上边（没有气泡时传头顶线），留一条缝接着叠。
    static func origin(size: NSSize, petFrame pet: NSRect, bubbleTop: CGFloat, visible: NSRect) -> NSPoint {
        var x = (pet.midX - size.width / 2).rounded()
        x = min(max(x, visible.minX + 4), visible.maxX - size.width - 4)
        var y = (bubbleTop + CardLook.stackGap).rounded()
        y = min(max(y, visible.minY + 4), visible.maxY - size.height - 4)
        return NSPoint(x: x, y: y)
    }

    func hit(at point: NSPoint, in bounds: NSRect) -> Hit? {
        for (i, rect) in cardRects.enumerated() {
            let r = rect.offsetBy(dx: bounds.minX, dy: bounds.minY)
            guard r.contains(point) else { continue }
            return Self.closeRect(r).contains(point) ? .dismiss(i) : .card(i)
        }
        // 只认画出来的那颗胶囊；两侧的透明边照旧穿透过去。
        if let a = autoRect, Self.pill(a.offsetBy(dx: bounds.minX, dy: bounds.minY), inset: Self.autoInset).contains(point) {
            return .auto
        }
        if let p = pillRect, Self.pill(p.offsetBy(dx: bounds.minX, dy: bounds.minY), inset: Self.pillInset).contains(point) {
            return .expand
        }
        return nil
    }

    func draw(in bounds: NSRect) {
        for (i, card) in shown.enumerated() {
            draw(card, in: cardRects[i].offsetBy(dx: bounds.minX, dy: bounds.minY))
        }
        if let a = autoRect?.offsetBy(dx: bounds.minX, dy: bounds.minY) {
            drawRow(tr("Auto", "自动"), in: a, inset: Self.autoInset)
        }
        if let p = pillRect?.offsetBy(dx: bounds.minX, dy: bounds.minY) {
            drawRow(expanded ? tr("Collapse", "收起") : tr("\(hidden) more", "还有 \(hidden) 条"), in: p, inset: Self.pillInset)
        }
    }

    /// 一条胶囊的实际矩形（画与命中共用）。
    private static func pill(_ rect: NSRect, inset: CGFloat) -> NSRect {
        NSRect(x: rect.minX + inset, y: rect.minY + 1, width: rect.width - 2 * inset, height: rect.height - 2)
    }

    private func drawRow(_ text: String, in rect: NSRect, inset: CGFloat) {
        let pill = Self.pill(rect, inset: inset)
        CardLook.box(pill, radius: pill.height / 2)
        let h = CardLook.lineHeight(CardLook.titleFont)
        CardLook.draw(text, CardLook.titleFont, CardLook.muted,
                      in: NSRect(x: pill.minX, y: pill.midY - h / 2, width: pill.width, height: h), center: true)
    }

    /// 自查用：给出几个代表点（每张卡的正文、✕，自动那条，收起那条）。
    static func probePoints(_ layout: CardStackLayout, in bounds: NSRect) -> [(String, NSPoint)] {
        var out: [(String, NSPoint)] = []
        for (i, rect) in layout.cardRects.enumerated() {
            let r = rect.offsetBy(dx: bounds.minX, dy: bounds.minY)
            out.append(("第\(i)张正文", NSPoint(x: r.midX, y: r.midY)))
            out.append(("第\(i)张的✕", NSPoint(x: closeRect(r).midX, y: closeRect(r).midY)))
        }
        if let a = layout.autoRect { out.append(("自动", NSPoint(x: a.midX, y: a.midY))) }
        if let p = layout.pillRect { out.append(("收起/还有N条", NSPoint(x: p.midX, y: p.midY))) }
        out.append(("空白处", NSPoint(x: bounds.minX + 2, y: bounds.maxY - 1)))
        return out
    }

    private static func closeRect(_ card: NSRect) -> NSRect {
        NSRect(x: card.maxX - closeBox - 1, y: card.maxY - closeBox - 1, width: closeBox, height: closeBox)
    }

    private func draw(_ card: ActivityCard, in rect: NSRect) {
        CardLook.box(rect)
        // 左边一条状态色：橙＝等你回答，红＝出错，绿＝答完了，蓝＝在跑。
        let bar = NSBezierPath(roundedRect: NSRect(x: rect.minX + 4, y: rect.minY + 5,
                                                   width: Self.stripe, height: rect.height - 10),
                               xRadius: Self.stripe / 2, yRadius: Self.stripe / 2)
        Self.color(for: card.status).setFill()
        bar.fill()

        let x = rect.minX + 4 + Self.stripe + Self.stripeGap
        let right = rect.maxX - CardLook.padX
        var top = rect.maxY - CardLook.padY

        // 上行：聊天名（和气泡上行一样的淡色小字），右边让出 ✕。
        let th = CardLook.lineHeight(CardLook.titleFont)
        top -= th
        CardLook.draw(CardLook.singleLine(card.title), CardLook.titleFont, CardLook.muted,
                      in: NSRect(x: x, y: top, width: max(10, right - x - Self.closeBox), height: th))
        drawClose(in: Self.closeRect(rect))

        // 下行：在做什么（和气泡下行一样的深色粗字），右边一个短状态标签。
        top -= CardLook.lineGap
        let ch = CardLook.lineHeight(CardLook.currentFont)
        top -= ch
        let label = card.status.label
        let labelW = ceil(CardLook.width(label, CardLook.smallFont)) + 1
        let lh = CardLook.lineHeight(CardLook.smallFont)
        CardLook.draw(label, CardLook.smallFont, Self.color(for: card.status).shadow(withLevel: 0.25) ?? CardLook.muted,
                      in: NSRect(x: right - labelW, y: top + (ch - lh) / 2, width: labelW, height: lh))
        CardLook.draw(CardLook.singleLine(card.subtitle), CardLook.currentFont, CardLook.ink,
                      in: NSRect(x: x, y: top, width: max(10, right - x - labelW - 6), height: ch))
    }

    private func drawClose(in box: NSRect) {
        let cross = NSBezierPath()
        let inset: CGFloat = 5
        cross.move(to: NSPoint(x: box.minX + inset, y: box.minY + inset))
        cross.line(to: NSPoint(x: box.maxX - inset, y: box.maxY - inset))
        cross.move(to: NSPoint(x: box.minX + inset, y: box.maxY - inset))
        cross.line(to: NSPoint(x: box.maxX - inset, y: box.minY + inset))
        cross.lineWidth = 1.2
        cross.lineCapStyle = .round
        CardLook.muted.withAlphaComponent(0.7).setStroke()
        cross.stroke()
    }
}

extension Array {
    /// 这摞卡的内容随时在变，按下标取时宁可拿到 nil 也不要越界。
    subscript(safe index: Int) -> Element? {
        indices.contains(index) ? self[index] : nil
    }
}

File: YukioWin/yukio/cardstack.py
"""头顶那摞卡片里“别的聊天”那几张：一条聊天一张，压着气泡往上叠，点一张就去那条聊天。

外观就是头顶气泡那张卡（`bubble.py` 里的那套圆角、描边、字号和两行排版），
只多了一条状态色和一个 ✕。用 Pillow 画成一张 RGBA 图交给分层窗口显示；
命中测试按卡片矩形算，卡与卡之间的缝隙不接收点击。

移植自 YukioPlayer/Sources/YukioPlayer/CardStack.swift，尺寸与排版一致。
"""

from __future__ import annotations

from typing import List, NamedTuple, Optional, Tuple

from PIL import Image, ImageDraw

from .bubble import (ACCENT, INK, MUTED, PAD_X, PAD_Y, PAPER, LINE_GAP, MAX_WIDTH, RADIUS,
                     CURRENT_SIZE, TITLE_SIZE, PROGRESS_SIZE, _fit, _line_height, _single_line,
                     _text_width, load_font)
from .cards import ActivityCard, CardStatus
from .l10n import tr

#: 和气泡同宽，叠起来是一摞齐的。
WIDTH = MAX_WIDTH
#: 叠起来时两张卡之间的缝。
STACK_GAP = 4.0
#: 收起时最多显示几张，其余折进“还有 N 条”。
COLLAPSED_COUNT = 3
STRIPE = 3.0
STRIPE_GAP = 5.0
CLOSE_BOX = 15.0
MORE_HEIGHT = 18.0

#: 左边那条状态色：橙＝等你回答，红＝出错，绿＝答完了，蓝＝在跑。
STATUS_COLOR = {
    CardStatus.waiting: (237, 158, 41),
    CardStatus.failed: (214, 79, 74),
    CardStatus.ready: (56, 173, 112),
    CardStatus.running: ACCENT,
}


def _darker(rgb, k: float = 0.75):
    return tuple(int(v * k) for v in rgb)


class Hit(NamedTuple):
    """点在了这摞卡的什么地方。"""

    #: "card" 打开那条聊天、"dismiss" 只收起这张、"expand" 展开／收起。
    kind: str
    #: 第几张卡（expand 时是 -1）。
    index: int


class CardStackLayout:
    """这摞卡的排版与绘制，与窗口无关（--cards 快照也用它）。

    scale：一个点等于多少像素（雪绪的缩放 × 屏幕缩放）。
    """

    def __init__(self, cards: List[ActivityCard], expanded: bool = False, scale: float = 1.0):
        self.scale = max(0.5, float(scale))
        self.expanded = bool(expanded)
        limit = len(cards) if expanded else min(COLLAPSED_COUNT, len(cards))
        #: 这摞里显示出来的卡（收起时是前几张）。
        self.shown: List[ActivityCard] = list(cards[:limit])
        #: 折进“还有 N 条”的张数。
        self.hidden = len(cards) - len(self.shown)

        s = self.scale
        self._title_font = load_font(TITLE_SIZE * s)
        self._current_font = load_font(CURRENT_SIZE * s)
        self._small_font = load_font(PROGRESS_SIZE * s)
        card_h = (2 * PAD_Y * s + _line_height(self._title_font) + LINE_GAP * s
                  + _line_height(self._current_font))

        #: 每张卡的矩形（左上原点，第 0 张在最下面、离气泡最近）。
        width = WIDTH * s
        rects: List[Tuple[float, float, float, float]] = []
        total = len(self.shown) * card_h + max(0, len(self.shown) - 1) * STACK_GAP * s
        if self.hidden > 0 or self.expanded:
            total += MORE_HEIGHT * s + STACK_GAP * s
        # 最要紧的那张挨着气泡（也就是最下面一张），所以从下往上排。
        y = total
        for _ in self.shown:
            y -= card_h
            rects.append((0.0, y, width, card_h))
            y -= STACK_GAP * s
        self._card_rects = rects
        self._more_rect = (0.0, 0.0, width, MORE_HEIGHT * s) if (self.hidden > 0 or self.expanded) else None

        self.size_px = (int(round(width)), int(round(max(0.0, total))))
        self.size_pt = (self.size_px[0] / s, self.size_px[1] / s)

    # MARK: 摆位与命中

    @staticmethod
    def origin(size_pt: Tuple[float, float], pet_rect: Tuple[float, float, float, float],
               bubble_top: float) -> Tuple[float, float]:
        """这摞卡的左上角（屏幕坐标，y 向下）：和气泡同一竖线上居中，压在气泡上面往上长。

        bubble_top 是气泡的上边（没有气泡时传头顶线），留一条缝接着叠。
        """
        x, y, w, h = pet_rect
        bx = round(x + w / 2 - size_pt[0] / 2)
        by = round(bubble_top - STACK_GAP - size_pt[1])
        return (bx, by)

    def hit(self, x: float, y: float) -> Optional[Hit]:
        """窗口内坐标（左上原点）上有没有可点的东西。"""
        for i, (rx, ry, rw, rh) in enumerate(self._card_rects):
            if rx <= x < rx + rw and ry <= y < ry + rh:
                cx0, cy0, cx1, cy1 = self._close_rect((rx, ry, rw, rh))
                if cx0 <= x < cx1 and cy0 <= y < cy1:
                    return Hit("dismiss", i)
                return Hit("card", i)
        if self._more_rect:
            rx, ry, rw, rh = self._more_rect
            if rx <= x < rx + rw and ry <= y < ry + rh:
                return Hit("expand", -1)
        return None

    def _close_rect(self, card):
        rx, ry, rw, rh = card
        s = self.scale
        box = CLOSE_BOX * s
        return (rx + rw - box - s, ry + s, rx + rw - s, ry + box + s)

    # MARK: 绘制

    def render(self) -> Image.Image:
        s = self.scale
        image = Image.new("RGBA", self.size_px, (0, 0, 0, 0))
        draw = ImageDraw.Draw(image)
        for card, rect in zip(self.shown, self._card_rects):
            self._draw_card(draw, card, rect)
        if self._more_rect:
            rx, ry, rw, rh = self._more_rect
            pill = (rx + 30 * s, ry + s, rx + rw - 30 * s, ry + rh - s)
            draw.rounded_rectangle(pill, radius=(pill[3] - pill[1]) / 2, fill=PAPER,
                                   outline=INK + (41,), width=max(1, int(round(s))))
            text = tr("Collapse", "收起") if self.expanded else tr("%d more", "还有 %d 条") % self.hidden
            tw = _text_width(draw, text, self._title_font)
            th = _line_height(self._title_font)
            draw.text(((pill[0] + pill[2]) / 2 - tw / 2, (pill[1] + pill[3]) / 2 - th / 2),
                      text, font=self._title_font, fill=MUTED + (255,))
        return image

    def _draw_card(self, draw, card: ActivityCard, rect) -> None:
        s = self.scale
        rx, ry, rw, rh = rect
        draw.rounded_rectangle((rx, ry, rx + rw - 1, ry + rh - 1), radius=RADIUS * s, fill=PAPER,
                               outline=INK + (41,), width=max(1, int(round(s))))
        color = STATUS_COLOR.get(card.status, ACCENT)
        bar = (rx + 4 * s, ry + 5 * s, rx + 4 * s + STRIPE * s, ry + rh - 5 * s)
        draw.rounded_rectangle(bar, radius=STRIPE * s / 2, fill=color + (255,))

        x = rx + (4 + STRIPE + STRIPE_GAP) * s
        right = rx + rw - PAD_X * s
        top = ry + PAD_Y * s

        # 上行：聊天名（和气泡上行一样的淡色小字），右边让出 ✕。
        th = _line_height(self._title_font)
        draw.text((x, top), _fit(draw, _single_line(card.title), self._title_font,
                                 max(10.0, right - x - CLOSE_BOX * s)),
                  font=self._title_font, fill=MUTED + (255,))
        self._draw_close(draw, self._close_rect(rect))
        top += th + LINE_GAP * s

        # 下行：在做什么（和气泡下行一样的深色粗字），右边一个短状态标签。
        ch = _line_height(self._current_font)
        label = card.status_label
        label_w = _text_width(draw, label, self._small_font) + 1
        lh = _line_height(self._small_font)
        draw.text((right - label_w, top + (ch - lh) / 2), label,
                  font=self._small_font, fill=_darker(color) + (255,))
        draw.text((x, top), _fit(draw, _single_line(card.subtitle), self._current_font,
                                 max(10.0, right - x - label_w - 6 * s)),
                  font=self._current_font, fill=INK + (255,))

    def _draw_close(self, draw, box) -> None:
        x0, y0, x1, y1 = box
        inset = 5 * self.scale
        width = max(1, int(round(1.2 * self.scale)))
        color = MUTED + (178,)
        draw.line((x0 + inset, y0 + inset, x1 - inset, y1 - inset), fill=color, width=width)
        draw.line((x0 + inset, y1 - inset, x1 - inset, y0 + inset), fill=color, width=width)
